src/backtracking/n_queens.rs

```diff
@@ -1,81 +1,108 @@
-
-    nqueens_solver(4) => Returns two solutions:
-    Solution 1:
-    [
-        ".Q..",
-        "...Q",
-        "Q...",
-        "..Q."
-    ]
-
-    Solution 2:
-    [
-        "..Q.",
-        "Q...",
-        "...Q",
-        ".Q.."
-    ]
-*/
-
+//! This module provides functionality to solve the N-Queens problem.
+//!
+//! The N-Queens problem is a classic chessboard puzzle where the goal is to
+//! place N queens on an NxN chessboard so that no two queens threaten each
+//! other. Queens can attack each other if they share the same row, column, or
+//! diagonal.
+//!
+//! This implementation solves the N-Queens problem using a backtracking algorithm.
+//! It starts with an empty chessboard and iteratively tries to place queens in
+//! different rows, ensuring they do not conflict with each other. If a valid
+//! solution is found, it's added to the list of solutions.
+
+/// Solves the N-Queens problem for a given size and returns a vector of solutions.
+///
+/// # Arguments
+///
+/// * `n` - The size of the chessboard (NxN).
+///
+/// # Returns
+///
+/// A vector containing all solutions to the N-Queens problem.
 pub fn n_queens_solver(n: usize) -> Vec<Vec<String>> {
-    let mut board = vec![vec!['.'; n]; n];
-    let mut solutions = Vec::new();
-    solve(&mut board, 0, &mut solutions);
-    solutions
+    let mut solver = NQueensSolver::new(n);
+    solver.solve()
+}
+
+/// Represents a solver for the N-Queens problem.
+struct NQueensSolver {
+    // The size of the chessboard
+    size: usize,
+    // A 2D vector representing the chessboard where '.' denotes an empty space and 'Q' denotes a queen
+    board: Vec<Vec<char>>,
+    // A vector to store all valid solutions
+    solutions: Vec<Vec<String>>,
 }
 
-fn is_safe(board: &[Vec<char>], row: usize, col: usize) -> bool {
-    // Check if there is a queen in the same column
-    for (i, _) in board.iter().take(row).enumerate() {
-        if board[i][col] == 'Q' {
-            return false;
+impl NQueensSolver {
+    /// Creates a new `NQueensSolver` instance with the given size.
+    ///
+    /// # Arguments
+    ///
+    /// * `size` - The size of the chessboard (N×N).
+    ///
+    /// # Returns
+    ///
+    /// A new `NQueensSolver` instance.
+    fn new(size: usize) -> Self {
+        NQueensSolver {
+            size,
+            board: vec![vec!['.'; size]; size],
+            solutions: Vec::new(),
         }
     }
 
-    // Check if there is a queen in the left upper diagonal
-    for i in (0..row).rev() {
-        let j = col as isize - (row as isize - i as isize);
-        if j >= 0 && board[i][j as usize] == 'Q' {
-            return false;
-        }
+    /// Solves the N-Queens problem and returns a vector of solutions.
+    ///
+    /// # Returns
+    ///
+    /// A vector containing all solutions to the N-Queens problem.
+    fn solve(&mut self) -> Vec<Vec<String>> {
+        self.solve_helper(0);
+        self.solutions.clone()
```

Comment:
```suggestion
        std::mem::take(&mut self.solutions)
```

Reply:
It's better than `clone` the entire solution. Good point.